### DISCUSSION TITLE
50+ AI Performance

### DISCUSSION THREAD

**Discussion by Jus2Cat** [Sixze/ALS-Refactored#118] [Q&A] [answered]
Hello. Thank you for your work. I more like your refactored version but I need many AI in my project and their performance is very important. I tested your version and ALSCommunity latest version. For testing was used monolithic anim bp. And there are my results with ryzen 5800h and rtx 3070 laptop:

ALSRefactored:

30AI - camera turned to AI ~12-13 mc, ~70-77 fps
	   camera turned to wall ~12 mc, ~80 fps

50AI - camera turned to AI ~19-20 mc, ~50-55 fps

ALSCommunity:

30AI - camera turned to wall  ~9 mc, 110 fps.
	 - camera turned to AI  ~10+ mc,  95 -100 fps
	 - run through crowd ~100 fps,  ~10 mc

50AI -  camera turned to wall  11-13mc , ~80 fps
	camera turned to AI 12-13mc, ~70-80 fps
	run through crowd 13-14mc, 70 fps

I'm not good in profiling but you can see Crowd with 50AI -> 55 vs 80 fps.
I tried to improve this result and I created another AI blueprint class from ALSBase c++ class. It gives me +5 fps.
I tried to find a difference with ALSCommunity character class defaults. And I found disabled checkbox "Enable update rate optimization" . I enabled it and It gave me same fps results with ALSCommunity. It even was a little bit better.
But If this checkbox enabled It breaks AI rotation (it has a jitters).

Does there a another way to improve crowd performance ?

**Comment by Sixze** [comment, marked as answer]
ALS was not designed with crowds in mind. Basically, most of the performance problems are related to the high complexity of the animation blueprint itself, and in order to improve performance, you will either have to simplify its components, or get rid of some features that you can do without.

Update Rate Optimizations is fully supported by my version of ALS, and the illusion of jitter was due to the fact that the animation blueprint is updated not every frame, but once every few frames. You can verify that the rotations are ok using the `a.URO.ForceAnimRate 10` console command.

There are also a couple of interesting resources about optimization, though I haven't studied them much, but I think they will come in handy:

https://docs.unrealengine.com/5.0/en-US/animation-budget-allocator-in-unreal-engine/
https://docs.unrealengine.com/5.0/en-US/animation-optimization-in-unreal-engine/
https://www.youtube.com/watch?v=CxHrNGa2KDE
